1.

In [ ]:
from IPython.display import display
%run Machine_Learning_Review_Classification.ipynb

3.11.11 (main, Dec  4 2024, 08:55:07) [GCC 11.4.0]
Pandas version: 2.2.2
Matplotlib version: 3.10.0
Numpy version: 2.0.2
SciKitLearn version: 1.6.1
My working directory:
/content
My new working directory:
/content
              precision    recall  f1-score   support

         0.0       0.76      0.85      0.81       970
         1.0       0.66      0.52      0.58       530

    accuracy                           0.73      1500
   macro avg       0.71      0.68      0.69      1500
weighted avg       0.73      0.73      0.72      1500


Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 

Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 
Null                   0.6467   0.608 
              precision    recall  f1-score   support

         0.0       0.76      0.85      0.80       970
         1.0       0.65      0.52      0.58       530

    accuracy  

In [ ]:
get_results(result_scores)


Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 
Null                   0.6467   0.608 
Logistic_L1_C_1        0.732    0.716 
Logistic_L1_C_01       0.726    0.706 
Logistic_L1_C_10       0.7347   0.718 
Logistic_L1_C_auto     0.7233   0.708 
Logistic_SL1_C_auto    0.7307   0.714 
RandomForest_noCV      0.9993   0.684 
RandomForest_CV        0.9953   0.692 
RandomForest_CV2       0.724    0.7   


Among all the classification models evaluated, Logistic_L1_C_10 delivered the strongest performance overall. It attained a training accuracy of 0.7347 and a test accuracy of 0.718, the highest among the models assessed. These results indicate that the model effectively balances learning from the training data while maintaining strong generalization to new, unseen data.

2.


In [ ]:
patient_df = df_patient.copy()

In [ ]:
patient_df["Gender"] = patient_df["Gender"].map({'male': 0, 'female': 1})
patient_df["Race"] = patient_df["Race"].map({'white':0, 'hispanic':1, 'black':2, 'other':3})

In [ ]:
X = patient_df[list(vars_left)]
y= patient_df['mortality']
X_train, X_test, y_train, y_test = train_test_split(X, np.ravel(y), random_state=20, test_size=0.20)

In [ ]:
X_train.shape

(16000, 25)

In [ ]:
X_test.shape

(4000, 25)

In [ ]:
y_train.shape

(16000,)

In [ ]:
y_test.shape

(4000,)

In [ ]:
import time
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Dictionary to store results
result_scores = {}

# Function to create and evaluate a logistic regression model with a specific solver
def create_linear_model(solver_name):
    model = LogisticRegression(
        fit_intercept=True,
        penalty=None,
        solver=solver_name,
        max_iter=10000
    )

    start_time = time.time()
    model.fit(X_train, y_train)
    end_time = time.time()

    train_accuracy = accuracy_score(y_train, model.predict(X_train))
    test_accuracy = accuracy_score(y_test, model.predict(X_test))
    elapsed_time = end_time - start_time

    result_scores[solver_name] = (train_accuracy, test_accuracy, elapsed_time)

# Function to display results
def print_results():
    header = f"\n{'Solver':20} {'Training Subset Accuracy':25} {'Test Subset Accuracy':25} {'Time (s)':10}"
    print(header)
    print("-" * len(header))
    for solver, (train_acc, test_acc, duration) in result_scores.items():
        print(f"{solver:20} {train_acc:<25.4f} {test_acc:<25.4f} {duration:<10.4f}")

# Run models with different solvers
solvers = ["lbfgs", "newton-cg", "newton-cholesky", "sag", "saga"]
for solver in solvers:
    create_linear_model(solver)

# Display the results
print_results()



Solver               Training Subset Accuracy  Test Subset Accuracy      Time (s)  
------------------------------------------------------------------------------------
lbfgs                0.7489                    0.7355                    0.2861    
newton-cg            0.7487                    0.7355                    0.0977    
newton-cholesky      0.7487                    0.7355                    0.0336    
sag                  0.7488                    0.7358                    2.7716    
saga                 0.7488                    0.7358                    5.4226    


3. Compare the results of the models in terms of their accuracy (use this as the performance metric to assess generalizability error on the holdout subset) and the time taken (use appropriate timing function). Summarize your results via a table with the following structure:

In [ ]:
print_results()


Solver               Training Subset Accuracy  Test Subset Accuracy      Time (s)  
------------------------------------------------------------------------------------
lbfgs                0.7489                    0.7355                    0.2861    
newton-cg            0.7487                    0.7355                    0.0977    
newton-cholesky      0.7487                    0.7355                    0.0336    
sag                  0.7488                    0.7358                    2.7716    
saga                 0.7488                    0.7358                    5.4226    


4. Among the solvers tested, newton-cholesky delivered the best overall performance. Although all solvers produced very similar accuracy on both the training and test sets, the variation in execution time was notable. newton-cholesky stood out by offering one of the top test accuracies (0.7355) while also being the fastest, completing the task in only 0.0336 seconds. While sag and saga achieved a slightly higher test accuracy (0.7358), they required significantly more time to execute. The final ranking prioritized both generalization performance and computational efficiency, with greater weight placed on the model's ability to generalize well and run quickly.